From 21399198c2c975b426c01777d2e4929b45ab2252 Mon Sep 17 00:00:00 2001
From: Yoni Izutskiver <yizutskiver@habana.ai>
Date: Wed, 16 Feb 2022 17:39:58 +0200
Subject: [PATCH] [SW-74686] gaudi2, greco: pci: mask ca aer

This WA prevents Intel hosts from crashing on continuous reads
that request more than 4B from the MSIx table region.

Change-Id: I039afa0bb09a5d128dbca1c7bcfc26daff932059
Signed-off-by: Yoni Izutskiver <yizutskiver@habana.ai>
---

diff --git a/arch/gaudi/arm/common.c b/arch/gaudi/arm/common.c
index 10e49e0..ce5bae9 100644
--- a/arch/gaudi/arm/common.c
+++ b/arch/gaudi/arm/common.c
@@ -669,6 +669,7 @@
 	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
 	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
 	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.dbi_uncorr_aer_mask = mmPCIE_DBI_UNCORR_ERR_MASK_OFF;
 	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
 	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
 	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 4a0b3eb..7c9286c 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -303,6 +303,7 @@
 	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
 }
 
+
 void preboot_disable_pcie_ltr(void)
 {
 	u32 cap_hdr_cfg, dev_cap2_reg;
@@ -558,6 +559,9 @@
 
 	/* set MSIx PBA offset */
 	set_pci_msix_pba_offset_val();
+
+	/* Mask Completer Abort AER */
+	hl_pcie_mask_ca_aer();
 
 	/* Enable and Mask LTSSM (erratum 102) */
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
@@ -1244,6 +1248,7 @@
 	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
 	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
 	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.dbi_uncorr_aer_mask = mmPCIE_DBI_UNCORR_ERR_MASK_OFF;
 	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
 	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
 	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index f419d61..f2dcabc 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -200,6 +200,7 @@
 	pci.dbi_bar2_reg = mmD0_PCIE_DBI_SIG_BAR2_REG;
 	pci.dbi_bar4_reg = mmD0_PCIE_DBI_SIG_BAR4_REG;
 	pci.dbi_status_command_reg = mmD0_PCIE_DBI_SIG_STATUS_COMMAND_REG;
+	pci.dbi_uncorr_aer_mask = mmD0_PCIE_DBI_SIG_UNCORR_ERR_MASK_OFF;
 	pci.aux_dbi_reg = mmD0_PCIE_AUX_DBI;
 	pci.aux_mem_space_en_reg = mmD0_PCIE_AUX_MEM_SPACE_EN;
 	pci.aux_flr_ctrl_reg = mmD0_PCIE_AUX_FLR_CTRL;
diff --git a/arch/goya/arm/common.c b/arch/goya/arm/common.c
index 12516b5..1d6d07c 100644
--- a/arch/goya/arm/common.c
+++ b/arch/goya/arm/common.c
@@ -328,6 +328,7 @@
 	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
 	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
 	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.dbi_uncorr_aer_mask = mmPCIE_DBI_UNCORR_ERR_MASK_OFF;
 	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
 	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
 	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
diff --git a/arch/greco/common/common.c b/arch/greco/common/common.c
index 051b41d..5516a59 100644
--- a/arch/greco/common/common.c
+++ b/arch/greco/common/common.c
@@ -556,6 +556,9 @@
 	/* PCIe disable LTR */
 	preboot_disable_pcie_ltr();
 
+	/* Mask Completer Abort AER */
+	hl_pcie_mask_ca_aer();
+
 	/* Enable and Mask LTSSM (erratum 102) */
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
@@ -650,6 +653,7 @@
 	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
 	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
 	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.dbi_uncorr_aer_mask = mmPCIE_DBI_UNCORR_ERR_MASK_OFF;
 	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
 	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
 	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
diff --git a/include/device.h b/include/device.h
index 23a73f4..e100e12 100644
--- a/include/device.h
+++ b/include/device.h
@@ -25,6 +25,7 @@
 	u32 dbi_bar2_reg;
 	u32 dbi_bar4_reg;
 	u32 dbi_status_command_reg;
+	u32 dbi_uncorr_aer_mask;
 	u32 aux_dbi_reg;
 	u32 aux_mem_space_en_reg;
 	u32 aux_flr_ctrl_reg;
diff --git a/lib/common/pci.c b/lib/common/pci.c
index a821987..26bdb80 100644
--- a/lib/common/pci.c
+++ b/lib/common/pci.c
@@ -224,3 +224,14 @@
 	reg |= BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
 	regmap_writel(reg, pci.dbi_status_command_reg);
 }
+
+#define AER_UNCORR_ERR_MASK_CMPLT_ABORT_ERR_SHIFT 15
+
+void hl_pcie_mask_ca_aer(void)
+{
+	u32 uncorr_aer_mask_reg_val;
+
+	uncorr_aer_mask_reg_val = regmap_readl(pci.dbi_uncorr_aer_mask);
+	uncorr_aer_mask_reg_val |= BIT(AER_UNCORR_ERR_MASK_CMPLT_ABORT_ERR_SHIFT);
+	regmap_writel(uncorr_aer_mask_reg_val, pci.dbi_uncorr_aer_mask);
+}
diff --git a/lib/common/pci.h b/lib/common/pci.h
index e26eefb..2111a03 100644
--- a/lib/common/pci.h
+++ b/lib/common/pci.h
@@ -21,7 +21,7 @@
 u32 hl_pci_get_ven_id_dev_id(void);
 void hl_pcie_disable_mem_access(void);
 void hl_pcie_enable_mem_access(void);
-
+void hl_pcie_mask_ca_aer(void);
 void hl_pcie_init(void);
 
 #endif /* _PCI_H_ */
